src/TrieSet.mo

```diff
@@ -38,19 +74,34 @@ module {
 
   /// Test if two sets are equal.
   public func equal<T>(s1 : Set<T>, s2 : Set<T>, eq : (T, T) -> Bool) : Bool {
-    // XXX: Todo: use a smarter check
-    func unitEqual(_ : (), _ : ()) : Bool { true };
-    Trie.equalStructure<T, ()>(s1, s2, eq, unitEqual)
+    if (Trie.size(s1) != Trie.size(s2)) return false;
+    for (k in keys(s1)) {
+      if (Trie.find<T,()>(s2, k, eq) == null) {
+        return false;
+      }
+    };
+    return true;
   };
```

Comment:
So this function assumes that keys in `s1` and `s2` are hashed the same way?

If so, I think you can (after the size equality check) flatten both trees to arrays and compare elements pairwise to implement this check in linear time. Currently I think the algorithm runs in `O(size1 * log(size2))` as opposed to `O(size1 == size2)`.

Reply:
I'm not sure that works because I'm pretty sure the array ordering is arbitrary and could depend on the history of construction. I'll see if I can construct a counter-example, but the original  'equal' test was broken because it compared the Trie representations directly as trees which could have different lists
 representing same buckets

Reply:
> could depend on the history of construction.

Yes, that's true.  But only because of the association lists, which do not enforce an ordering constraint (maybe they could?).

Another valid way to do this which could be more efficient is to assume the Trie has a canonical shape (it does, by its invariants) and only the association lists could differ, by key order.  Doing things that way would be a bit more efficient, maybe?, but more verbose.

Reply:
Can we go with this implementation for now since at least it appears to be correct, if not optimal?

Reply:
Ah right I forgot the lists are arbitrary length now. The implementation does seem correct, will approve